User profiles

afrasier opened the pull request:
This PR implements the baseline user profile for the TalentMAP system.

Including @jseppi on the reviewer list as this is going to be the biggest chunk of code in the Sprint, so it might be good to review it now in this separate PR.

* Add `UserProfile` model which links to the default Django authorization `User` model, extending it to allow the storage of `favorite_positions`, `position_preferences`, and `language_qualifications`
* Added `create_user` management command to allow the creation of different users for development purposes
* Added basic authentication via Sessions and Expiring Tokens
* Added serializers to allow for the creation and updating of `UserProfile`, `Qualification` objects and a common mixin to support changing serializers based upon endpoint action
* Updated common serializer `PrefetchedSerializer` to pass overridden child fields on to children. For example, the `Position` object has a nested `Post` object. Prior to this change, one could not specify `post__description` in the field limitation, as it would not be passed to the child. This is now fixed.
* Add testing fixtures for `authorized_client` and `authorized_user` for use in testing authorization required endpoints
* Added tests for new endpoints

* Added the following endpoints (:lock: means the endpoint requires a valid session/token)
   * `/api/v1/accounts/` - Supports session logins
   * POST `/api/v1/accounts/token/` - Supports token logins (default expiration: 1 day)
   * GET `/api/v1/accounts/profile/` - :lock: - Displays the logged-in user's profile
   * PATCH `/api/v1/accounts/profile/` - :lock: - Updates the user profile with the data payload
   * PUT `/api/v1/language/qualifications/` -  :lock: - Creates a new language qualification
   * GET `/api/v1/position/preferred/` - :lock: - Lists all positions, filtered by user's stored preferences
   * GET `/api/v1/position/favorites/` - :lock: - Lists all of the current user's favorited positions

afrasier commented:
Looks like @burgwyn is having some problems with the Expiring Tokens library; so I'll be moving that functionality into our codebase proper.

jseppi commented on talentmap_api/common/serializers.py:
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
 
+                # If our parent serializer has field limitations on the child, pass them down
+                child_override_fields = []
+                child_override_exclude = []
```
any way you could test `PrefetchedSerializer` in isolation?

afrasier replied:
Yeah, I'll split it out into some functions and test those in isolation

jseppi commented on talentmap_api/user_profile/models.py:
```diff
@@ -0,0 +1,34 @@
+        "grade__code": "05"
+    }
+    '''
+    position_preferences = JSONField(default=dict, help_text="JSON object containing filters representing a user's position preferences")
```
Should these preferences be validated in any way before being passed to a query? Or maybe before being saved? Looks like anything can be stuffed in there currently.

afrasier replied:
Good point, I will add that

jseppi commented on talentmap_api/user_profile/models.py:
```diff
@@ -0,0 +1,34 @@
+    favorite_positions = models.ManyToManyField('position.Position', related_name='favorited_by_users', help_text="Positions which this user has designated as a favorite")
+
+
+@receiver(post_save, sender=User)
```
👍

afrasier commented on talentmap_api/common/serializers.py:
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
+                            child_overrides.append(LOOKUP_SEP.join(split_field[1:]))
+
+                    # If we have child overrides, attach them to the child's kwargs
+                    if len(child_overrides) > 0:
```
👍

afrasier commented on talentmap_api/common/serializers.py:
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
+                child_override_exclude = []
+
+                for pair in [(override_fields, "override_fields"), (override_exclude, "override_exclude")]:
+                    overrides = pair[0]
```
Yes, down when the list of overrides is passed into the kwargs

jseppi replied:
oh, whoops, missed it 👍

afrasier commented:
@jseppi Fixed the issues with the comments, split out some of the common serializer functionality and tested that in isolation. Added some tests to the profile endpoints to the generic endpoint tester (Adding my own functionality for custom recipes) and added validation/testing to the position preferences.

jseppi commented on talentmap_api/user_profile/serializers.py:
```diff
@@ -53,6 +56,20 @@ class Meta:
+            try:
+                Position.objects.filter(Q(**data.get('position_preferences')))
+            except:
+                raise ValidationError("Position preferences must be a valid set of position filters")
```
👍

jseppi commented on talentmap_api/user_profile/tests/test_user_profile_endpoints.py:
```diff
@@ -0,0 +1,68 @@
+    assert list(authorized_user.profile.favorite_positions.values_list("id", flat=True)) == [2]
+    assert list(authorized_user.profile.language_qualifications.values_list("id", flat=True)) == [2]
+
+    # Post a bad position preferences
```
👍
